File: CubeController/src/ServoManager.py
from .Servo import Servo
import time

class ServoManager:
    def __init__(self):
        self._boot_servo = Servo(27)
        self._BOOT_REST_POS = 1166
        self._BOOT_MID_POS = 1050
        self._BOOT_KICK_POS = 940
        self._is_boot_up = False

        self._holder_servo = Servo(25)
        self._HOLDER_POS = [500, 1200, 1930, 2540]
        self._holder_pos_index = 0

        self._handle_servo = Servo(12)
        self._HANDLE_OPEN_POS = 1166
        self._HANDLE_CLOSE_POS = 1920
        self.resetServos()

    def lowerBoot(self):
        self._is_boot_up = False
        self._boot_servo.write_us(self._BOOT_REST_POS)
        time.sleep(1)
    
    def bootKick(self):
        if self._is_boot_up:
            self.lowerBoot()
        self._boot_servo.write_us(self._BOOT_KICK_POS)
        time.sleep_ms(400)
        self._boot_servo.write_us(self._BOOT_MID_POS)
        time.sleep_ms(600)
        self._boot_servo.write_us(self._BOOT_KICK_POS)
        time.sleep_ms(400)
        self._boot_servo.write_us(self._BOOT_REST_POS)
        time.sleep(1)
    
    def bootHalfKick(self):
        if self._is_boot_up:
            self.lowerBoot()
        self._boot_servo.write_us(self._BOOT_KICK_POS)
        time.sleep_ms(400)
        self._boot_servo.write_us(self._BOOT_MID_POS)
        time.sleep_ms(600)
        self._boot_servo.write_us(self._BOOT_KICK_POS)
        time.sleep(1)
        self._is_boot_up = True
    
    def handleOpen(self):
        if self._is_boot_up:
            self.lowerBoot()
        self._handle_servo.write_us(self._HANDLE_OPEN_POS)
        time.sleep(1)
    
    def handleClose(self):
        if self._is_boot_up:
            self.lowerBoot()
        self._handle_servo.write_us(self._HANDLE_CLOSE_POS)
        time.sleep(1)
    
    def holderTurn(self, times: int):
        if self._is_boot_up:
            self.lowerBoot()
        self._holder_pos_index = (self._holder_pos_index + times) % len(self._HOLDER_POS)
        self._holder_servo.write_us(self._HOLDER_POS[self._holder_pos_index])
        time.sleep_ms(500)
    
    def resetServos(self):
        self._boot_servo.write_us(self._BOOT_REST_POS)
        self._is_boot_up = False
        self._handle_servo.write_us(self._HANDLE_OPEN_POS)
        time.sleep(1)
        self._holder_pos_index = 0
        self._holder_servo.write_us(self._HOLDER_POS[0])
        time.sleep_ms(500)

File: CubeController/src/ServoManager.cpp
#include "ServoManager.h"

Servo ServoManager::bootServo;
Servo ServoManager::handleServo;
Servo ServoManager::holderServo;
bool ServoManager::isBootUp;
int ServoManager::holderPosIndex;

void ServoManager::init() {
    isBootUp = false;
    holderPosIndex = 0;
    ESP32PWM::allocateTimer(0);
    bootServo.setPeriodHertz(50);
    bootServo.attach(BOOT_SERVO_PIN, SERVO_MIN, SERVO_MAX);
    handleServo.setPeriodHertz(50);
    handleServo.attach(HANDLE_SERVO_PIN, SERVO_MIN, SERVO_MAX);
    holderServo.setPeriodHertz(50);
    holderServo.attach(HOLDER_SERVO_PIN, SERVO_MIN, 2540);
    resetServos();
}

void ServoManager::lowerBoot() {
    isBootUp = false;
    bootServo.writeMicroseconds(bootRestPos);
    delay(1000);
}

void ServoManager::bootKick() {
    if(isBootUp) {
        lowerBoot();
    }
    bootServo.writeMicroseconds(bootKickPos);
    delay(400);
    bootServo.writeMicroseconds(bootMidPos);
    delay(600);
    bootServo.writeMicroseconds(bootKickPos);
    delay(400);
    bootServo.writeMicroseconds(bootRestPos);
    delay(1000);
}

void ServoManager::bootHalfKick() {
    if(isBootUp) {
        lowerBoot();
    }
    bootServo.writeMicroseconds(bootKickPos);
    delay(400);
    bootServo.writeMicroseconds(bootMidPos);
    delay(600);
    bootServo.writeMicroseconds(bootKickPos);
    delay(1000);
    isBootUp = true;
}

void ServoManager::handleOpen() {
    if(isBootUp) {
        lowerBoot();
    }
    handleServo.writeMicroseconds(handleOpenPos);
    delay(1000);
}

void ServoManager::handleClose() {
    if(isBootUp) {
        lowerBoot();
    }
    handleServo.writeMicroseconds(handleClosePos);
    delay(1000);
}

void ServoManager::holderTurn(int times) {
    if(isBootUp) {
        lowerBoot();
    }
    holderPosIndex += times;
    holderPosIndex %= 4;
    holderServo.writeMicroseconds(holderPos[holderPosIndex]);
    delay(1500);
}

void ServoManager::resetServos() {
    bootServo.writeMicroseconds(bootRestPos);
    handleServo.writeMicroseconds(handleOpenPos);
    delay(1000);
    holderPosIndex = 0;
    holderServo.writeMicroseconds(holderPos[0]);
    delay(500);
}